Begin implementing provisional navigation cancelling with site isolation

achristensen07 opened the pull request:
#### 025e1d2fda31b65326961d5c99039356b2eb7ece
<pre>
Begin implementing provisional navigation cancelling with site isolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=273971">https://bugs.webkit.org/show_bug.cgi?id=273971</a>
<a href="https://rdar.apple.com/127836397">rdar://127836397</a>

Reviewed by Charlie Wolfe.

When more than two processes are involved, there is an unavoidable time between when
the UI process tells one web process to start a provisional load and when it tells another
web process to start another provisional load that the first web process may have sent a
message to the UI process saying the provisional load has started.  If we don&apos;t ignore this
message, the UI process accounting of which frames are in a provisional load state and in
which process gets messed up.  In addition, when starting the second provisional load, we
need to tell the first provisional load process to destroy its provisional frame.

I renamed &quot;prepareForProvisionalNavigationInProcess&quot; to &quot;prepareForProvisionalLoadInProcess&quot;
because &quot;Navigation&quot; is a term we use mostly consistently to refer to a main frame load.

I renamed &quot;provisionalLoadFailed&quot; to &quot;destroyProvisionalFrame&quot; because it is now used for
more than just when a provisional load fails.  It&apos;s now used for when a provisional load
is intentionally cancelled.

This is just what it took to get the most controlled case of provisional load cancelling
working reliably with a test.  There&apos;s still more work to be done in this area, so I left
a FIXME comment in the test indicating how to test those cases.  I think we&apos;ll eventually
have to come up with something better than using Navigation::currentRequestIsCrossSiteRedirect.

* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
(WebKit::FrameLoadState::didFailProvisionalLoad):
(WebKit::FrameLoadState::didCommitLoad):
(WebKit::FrameLoadState::didFinishLoad):
(WebKit::FrameLoadState::didFailLoad):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::didReceiveMessage):
(WebKit::RemotePageProxy::didStartProvisionalLoadForFrame):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalNavigationInProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidStartProvisionalLoad):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, CancelProvisionalLoad)):

Canonical link: <a href="https://commits.webkit.org/278624@main">https://commits.webkit.org/278624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c338bbe7d95b0f54281969228d2f66f796258ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51120 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30422 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3453 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54377 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1810 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53423 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1484 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41635 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53219 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28058 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44060 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22751 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1316 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47363 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1388 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/56244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26230 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1282 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/56244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27478 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44127 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/56244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28359 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7431 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27209 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -127,6 +127,17 @@ std::unique_ptr<ProvisionalFrameProxy> WebFrameProxy::takeProvisionalFrame()
+    if (m_provisionalFrame)
+        return m_provisionalFrame->process();
+    if (isMainFrame()) {
+        if (WeakPtr provisionalPage = m_page ? m_page->provisionalPageProxy() : nullptr)
```
From the current call site, this will never be non-null, since we will clear `m_provisionalPage` before calling this function. But maybe this function could be used in other places in the future.

achristensen07 replied:
m_page is never null, but this addition of main frame provisional page proxies was needed to keep the window.open tests working.

charliewolfe replied:
`m_page->provisionalPageProxy()` would be null, not `m_page`

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4645,7 +4645,21 @@ void WebPageProxy::continueNavigationInNewProcess(API::Navigation& navigation, W
+            navigation = Ref { navigation },
+            previousProvisionalLoadProcess = Ref { frame.provisionalLoadProcess() }
+        ] () mutable {
+            if (frame->frameLoadState().state() == FrameLoadState::State::Provisional && !navigation->currentRequestIsCrossSiteRedirect()) {
```
Eventually it seems like we should also add tests to make sure things work right when provisional loads fail after a cross-site navigation.

achristensen07 replied:
Definitely.  This is a beginning.

achristensen07 committed:
Begin implementing provisional navigation cancelling with site isolation
https://bugs.webkit.org/show_bug.cgi?id=273971
rdar://127836397

Reviewed by Charlie Wolfe.

When more than two processes are involved, there is an unavoidable time between when
the UI process tells one web process to start a provisional load and when it tells another
web process to start another provisional load that the first web process may have sent a
message to the UI process saying the provisional load has started.  If we don't ignore this
message, the UI process accounting of which frames are in a provisional load state and in
which process gets messed up.  In addition, when starting the second provisional load, we
need to tell the first provisional load process to destroy its provisional frame.

I renamed "prepareForProvisionalNavigationInProcess" to "prepareForProvisionalLoadInProcess"
because "Navigation" is a term we use mostly consistently to refer to a main frame load.

I renamed "provisionalLoadFailed" to "destroyProvisionalFrame" because it is now used for
more than just when a provisional load fails.  It's now used for when a provisional load
is intentionally cancelled.

This is just what it took to get the most controlled case of provisional load cancelling
working reliably with a test.  There's still more work to be done in this area, so I left
a FIXME comment in the test indicating how to test those cases.  I think we'll eventually
have to come up with something better than using Navigation::currentRequestIsCrossSiteRedirect.

* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
(WebKit::FrameLoadState::didFailProvisionalLoad):
(WebKit::FrameLoadState::didCommitLoad):
(WebKit::FrameLoadState::didFinishLoad):
(WebKit::FrameLoadState::didFailLoad):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::didReceiveMessage):
(WebKit::RemotePageProxy::didStartProvisionalLoadForFrame):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalNavigationInProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidStartProvisionalLoad):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, CancelProvisionalLoad)):

Canonical link: https://commits.webkit.org/278624@main
